src/main/java/com/iexec/worker/executor/TaskExecutorService.java

```diff
@@ -212,24 +202,85 @@ private String compute(ContributionAuthorization contributionAuth) {
 
         if (isDatasetDecryptionNeeded && !isDatasetDecrypted) {
             customFeignClient.updateReplicateStatus(chainTaskId, COMPUTE_FAILED);
-            stdout = "Failed to decrypt dataset, URI:" + taskDescription.getDatasetUri();
-            log.error(stdout + " [chainTaskId:{}]", chainTaskId);
-            return stdout;
+            message = "Failed to decrypt dataset, URI:" + taskDescription.getDatasetUri();
+            log.error(message + " [chainTaskId:{}]", chainTaskId);
+            return message;
+        }
+
+        String error = checkContributionAbility(chainTaskId);
+        if (!error.isEmpty()) {
+            return error;
         }
 
         // compute
         String datasetFilename = datasetService.getDatasetFilename(taskDescription.getDatasetUri());
-        stdout = dockerComputationService.dockerRunAndGetLogs(taskDescription, datasetFilename);
+        message = dockerComputationService.dockerRunAndGetLogs(taskDescription, datasetFilename);
 
-        if (stdout.isEmpty()) {
+        if (message.isEmpty()) {
             customFeignClient.updateReplicateStatus(chainTaskId, COMPUTE_FAILED);
-            stdout = "Failed to start computation";
-            log.error(stdout + " [chainTaskId:{}]", chainTaskId);
-            return stdout;
+            message = "Failed to start computation";
+            log.error(message + " [chainTaskId:{}]", chainTaskId);
+            return message;
         }
 
         customFeignClient.updateReplicateStatus(chainTaskId, COMPUTED);
-        return stdout;
+        return message;
+    }
+
+    private String tryToDownloadApp(TaskDescription taskDescription) {
```

Comment:
Can we do something like
`getAppDownloadStatus(..)` https://github.com/iExecBlockchainComputing/iexec-worker/blob/c7d55c3f7219840197d88c2dc40d9d1ae7566c5f/src/main/java/com/iexec/worker/chain/ContributionService.java#L45
instead of setting inside/outside statuses ?

Reply:
(or `Pair getAppDownloadStatusAndMessage(..)` )

Reply:
You are right a big refactoring is needed for this entire method.